### PR TITLE
Introduce the entrypoint and sidecar containers

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Introduce the entrypoint and sidecar containers

In order to wrap the test execution, two new utilies are necessary. The
entrypoint container executes the test process and redirects the output
from `std{out,err}` to a file. When the test process has finished, the
entrypoint writes the test's exit code to a marker file. The sidecar
container waits and watches for the marker file to be present and, when
it is, generates `finished.json` from the exit code and uploads that,
the process log and any artifacts to GCS using the `$JOB_SPEC`.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Run hack.update-deps.sh to pull in fsnotify

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/area prow
/kind feature
/cc @fejta @spxtr 
/assign @kargakis @BenTheElder @cjwagner 

This is the final installment of the utilities. Next we can create `PodSpec`s to use them all.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Rebase from the outset? YUCK

File: Gopkg.lock [hunk 1]
```diff
@@ -191,6 +191,12 @@
   revision = "2dd44038f0b95ae693b266c5f87593b5d2fdd78d"
   version = "v2.5.0"
 
+[[projects]]
+  name = "github.com/fsnotify/fsnotify"
```
Comment (BenTheElder) [inline Gopkg.lock]: @cjwagner we should use this in the config agent to drive down config reload times :-)

Comment (BenTheElder) [reply Gopkg.lock]: I've been meaning to do this, and now somebody has vendored this for me :joy:

Comment (spxtr) [reply Gopkg.lock]: I'm always worried that it's going to miss some event somehow. With fs watchers there are always edge cases. Polling erases the worry :)

Comment (BenTheElder) [reply Gopkg.lock]: sure, but we can poll *with* notifying. we would *actually* miss some events only notifying. we've had problems with this with some kubernetes tests :-)

Comment (spxtr) [reply Gopkg.lock]: Oh yeah, polling + notifying is fine with me. Not sure how important it is to get <1 min config reload times tho.

File: prow/cmd/sidecar/main.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		defer watcher.Close()
+
+		group := sync.WaitGroup{}
+
+		group.Add(1)
+		go func() {
+			defer group.Done()
+			for {
+				select {
```
Comment (BenTheElder) [inline prow/cmd/sidecar/main.go]: we should probably also select on some low frequency ticker here to check for the file

File: Gopkg.lock (hunk 1 above)
Comment (BenTheElder) [reply Gopkg.lock]: frequently we update prow config to patch some misconfigured test and then someone sees the config update message on github and triggers a new test but the config hasn't actually been reloaded yet. fast reload times for job config would be nice, but definitely not important

Comment (stevekuznetsov) [reply Gopkg.lock]: I should add a poll to this loop ...

**Comment by stevekuznetsov** [comment]
Added a poll to the watch

File: prow/cmd/sidecar/main.go (hunk 2 above)
Comment (BenTheElder) [reply prow/cmd/sidecar/main.go]: this is done now, xref https://github.com/kubernetes/test-infra/pull/7058/files#r171614657 🙃

**Comment by stevekuznetsov** [comment]
Whoops, could have squashed that

**Pull request merged**